### PR TITLE
Florencia.ferraiolo/fragment dialog feature

### PR DESCRIPTION
**Pull request by florenciaferraiolo**
## Description
Add the Fragment Dialog feature with MVVM structure. The fragment shows an editText and two buttons, the "Confirm" one will send the string that the user puts in the editText to show in the MainActivity.

### Device 1:
NexusS1

NexusS_2

NexusS_3

NexusS_4

### Device 2:
Pixel1_1

Pixel1_2

Pixel1_3

Pixel1_4

## Tested on

Device 1: Nexus S
Device 2: Pixel 1

### CODE REVIEW

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    private fun setListeners() {
+        binding.confirmButton.setOnClickListener {
+            val updateParkingSpaces = binding.parkingAvailableEditText.text.toString()
+            viewModel.updateParkingSpacesValue(updateParkingSpaces)
+            listener.listenerParkingSpaces(updateParkingSpaces)
+            dismiss()
+        }
+        binding.cancelButton.setOnClickListener {
+            dismiss()
+        }
```
Comment (mauroLonghi) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: If you have more than one binding view used, you could use a scope function to updagrade your code. Like binding.apply , .run, or other

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.DialogFragment
+import com.mobile.tandil.kotlinbaseproject.databinding.FragmentMainDialogBinding
+import com.mobile.tandil.kotlinbaseproject.listener.ListenerSetParkingSpaces
+import com.mobile.tandil.kotlinbaseproject.mvvm.viewmodel.MainDialogViewModel
+import com.mobile.tandil.kotlinbaseproject.utils.Constants
+
+class MainDialogFragment() : DialogFragment() {
```
Comment (mauroLonghi) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: pls use a name more representative, to indicate in code whats the propose of this Dialog. Suggestion: ParkingSizeDialogFragmet.
Also, do the same to the ViewModel

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.mobile.tandil.kotlinbaseproject.listener
+
+import java.io.Serializable
+
+interface ListenerSetParkingSpaces : Serializable {
```
Comment (mauroLonghi) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt:5]: it's necesary defined to Serializable the interface?

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainModel.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.mobile.tandil.kotlinbaseproject.mvvm.model
+
+class MainModel {
+    private var parkingSpacesAvailable: String = ""
```
Comment (mauroLonghi) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainModel.kt]: not use hardcode pls

File: app/build.gradle [hunk 5]
```diff
@@ -24,6 +24,9 @@ android {
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+            create("customDebugType") {
```
Comment (SartiMau) [inline app/build.gradle]: what is this for?

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/activity/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        setContentView(binding.root)
+
+        viewModel.getValue().observe({ lifecycle }, ::updateUI)
+
+        setListeners()
+        viewModel.initialState()
+    }
+
+    private fun setListeners() {
+        binding.updateButton.setOnClickListener {
```
Comment (SartiMau) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/activity/MainActivity.kt:30]: this should call the viewModel, the viewModel set a new LiveData state and the activity react to that state

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainDialogModel.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.mobile.tandil.kotlinbaseproject.mvvm.model
+
+class MainDialogModel {
+    private var parkingSpaces = ""
```
Comment (SartiMau) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainDialogModel.kt]: do not hardcode, please us a constant for this

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainDialogViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package com.mobile.tandil.kotlinbaseproject.mvvm.viewmodel
+
+import androidx.lifecycle.ViewModel
+import com.mobile.tandil.kotlinbaseproject.mvvm.model.MainDialogModel
+
+class MainDialogViewModel : ViewModel() {
```
Comment (SartiMau) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainDialogViewModel.kt]: the model should be passed as parameter in the viewModel constructor

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        super.onViewCreated(view, savedInstanceState)
+        listener = arguments?.getSerializable(Constants.LISTENER_PARKING_SPACES_SETTING_KEY) as ListenerSetParkingSpaces
+        setListeners()
+    }
+
+    private fun setListeners() {
+        binding.confirmButton.setOnClickListener {
+            val updateParkingSpaces = binding.parkingAvailableEditText.text.toString()
+            viewModel.updateParkingSpacesValue(updateParkingSpaces)
+            listener.listenerParkingSpaces(updateParkingSpaces)
```
Comment (SartiMau) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: this is not following the arch. When the info is saved the viewModel should take the parking size from the model and set a new state on the liveData. Then the activity should react to that state, call this listener and close the dialog

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            dismiss()
+        }
+        binding.cancelButton.setOnClickListener {
+            dismiss()
```
Comment (SartiMau) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: please follow the arch

File: app/src/main/res/layout/fragment_main_dialog.xml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <EditText
+        android:id="@+id/parking_available_edit_text"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:inputType="text"
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_main_dialog.xml]: instead of text, I think that should be a number

File: app/src/main/res/layout/fragment_main_dialog.xml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    <Button
+        android:id="@+id/cancel_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="@string/cancel_button_text"
+        app:layout_constraintTop_toBottomOf="@id/parking_available_edit_text"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toLeftOf="@id/confirm_button"
+        app:layout_constraintHorizontal_chainStyle="spread"/>
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_main_dialog.xml]: do you know that is this property for?

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -1,5 +1,8 @@
 <resources>
     <string name="app_name">KotlinBaseProject</string>
-    <string name="welcome_text_field_text">Hola!</string>
-    <string name="check_button_text">OK</string>
+    <string name="parking_available_field_text">0</string>
```
Comment (SartiMau) [inline app/src/main/res/values/strings.xml]: this zero should be in the integers resource

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 1 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: the other Mau told me in the comments below to do it in a different way, can we have a call about it?

File: app/build.gradle (hunk 5 above)
Comment (florenciaferraiolo) [reply app/build.gradle]: i use it to run the debugger

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 9 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: the other Mau told me in the comments below to do it in a different way, can we have a call about it?

File: app/src/main/res/layout/fragment_main_dialog.xml (hunk 11 above)
Comment (florenciaferraiolo) [reply app/src/main/res/layout/fragment_main_dialog.xml]: im using as string to save the parking size, so i thought it would be better to use a text

File: app/src/main/res/layout/fragment_main_dialog.xml (hunk 12 above)
Comment (florenciaferraiolo) [reply app/src/main/res/layout/fragment_main_dialog.xml]: yes, it displays the buttons next to each other as horizontal chain with equals margins

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (florenciaferraiolo) [reply app/src/main/res/values/strings.xml]: but im using strings, should i change it anyway?

File: app/src/main/res/layout/fragment_main_dialog.xml (hunk 12 above)
Comment (SartiMau) [reply app/src/main/res/layout/fragment_main_dialog.xml]: spread is the default value so you do not need to put it

File: app/src/main/res/layout/fragment_main_dialog.xml (hunk 11 above)
Comment (SartiMau) [reply app/src/main/res/layout/fragment_main_dialog.xml]: the editText always will return a string but with this property you will set the keyboard type, and as you are waiting numbers, not letters, the best will be to configure the keyboard in that way

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (SartiMau) [reply app/src/main/res/values/strings.xml]: this will be always a sting but the zero should be in an integer resource

File: app/build.gradle (hunk 5 above)
Comment (SartiMau) [reply app/build.gradle]: It is really necessary? I never need it in the past and I can debug the code. Try removing it and see the behavior please

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 9 above)
Comment (SartiMau) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: sure, we can talk about it later, just open a chat and will see it

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 1 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 2 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainModel.kt (hunk 4 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainModel.kt]: done

File: app/build.gradle (hunk 5 above)
Comment (florenciaferraiolo) [reply app/build.gradle]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/activity/MainActivity.kt (hunk 6 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/activity/MainActivity.kt:30]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainDialogModel.kt (hunk 7 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/model/MainDialogModel.kt]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainDialogViewModel.kt (hunk 8 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainDialogViewModel.kt]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 9 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 10 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: done

File: app/src/main/res/layout/fragment_main_dialog.xml (hunk 11 above)
Comment (florenciaferraiolo) [reply app/src/main/res/layout/fragment_main_dialog.xml]: done

File: app/src/main/res/layout/fragment_main_dialog.xml (hunk 12 above)
Comment (florenciaferraiolo) [reply app/src/main/res/layout/fragment_main_dialog.xml]: done

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (florenciaferraiolo) [reply app/src/main/res/values/strings.xml]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 1 above)
Comment (mauroLonghi) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: I continue saw this issue

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt (hunk 3 above)
Comment (mauroLonghi) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt:5]: ?

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    data class ParkingData(
+        val state: CheckState
+    )
+
+    enum class CheckState {
+        INITIAL,
+        SHOW_FRAGMENT
+    }
```
Comment (mauroLonghi) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt]: this class must be outside the MainViewModel

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,37 @@
+package com.mobile.tandil.kotlinbaseproject.mvvm.viewmodel
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.mobile.tandil.kotlinbaseproject.mvvm.model.MainModel
+
+class MainViewModel : ViewModel() {
+
+    private val model = MainModel()
```
Comment (mauroLonghi) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt]: this Model must be at the constructor

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt (hunk 1 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/fragment/MainDialogFragment.kt]: It's changed in the new commit

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt (hunk 14 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt (hunk 15 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/MainViewModel.kt]: done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt (hunk 3 above)
Comment (mauroLonghi) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt:5]: why you are using this " : Serializable " ??

File: app/build.gradle [hunk 16]
```diff
@@ -37,11 +37,6 @@ android {
     lintOptions {
         abortOnError false
     }
-
-    buildFeatures {
```
Comment (SartiMau) [inline app/build.gradle:41]: Why did you delete this? Are you sure it is working?

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/ParkingSizeUpdateDialogViewModel.kt [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        mutableLiveData.value = ParkingSpaceData(UpdateState.CLOSE_DIALOG, Constants.INITIAL_SIZE_KEY)
+    }
+
+    data class ParkingSpaceData(
+        val state: UpdateState,
+        val size: String
+    )
+}
+
+enum class UpdateState {
```
Comment (SartiMau) [inline app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/ParkingSizeUpdateDialogViewModel.kt]: please, put the enum inside the viewModel, below the ParkingSpaceData

File: app/src/main/res/layout/fragment_parking_size_dialog.xml [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    <EditText
+        android:id="@+id/parking_available_edit_text"
+        android:layout_width="@dimen/parking_available_edit_text_width"
+        android:layout_height="@dimen/parking_available_edit_text_height"
+        android:inputType="number"
+        app:layout_constraintBottom_toTopOf="@id/confirm_button"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/fragment_tittle_text_view"
+        />
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_parking_size_dialog.xml]: please put this in the line above

File: app/src/main/res/values/dimen.xml [hunk 19]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <dimen name="parking_available_edit_text_width">250dp</dimen>
+    <dimen name="parking_available_edit_text_height">80dp</dimen>
+    <dimen name="fragment_tittle_text_view_text_size">16sp</dimen>
+    <dimen name="layout_marginBottom">20dp</dimen>
```
Comment (SartiMau) [inline app/src/main/res/values/dimen.xml]: please try to put a more representative name. remember the naming conventions https://docs.google.com/document/d/1kIL-jK55v_kWITs9qcW7d-0yJemUtLA74hQzm1_S-5k/edit?usp=sharing

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (SartiMau) [reply app/src/main/res/values/strings.xml]: this still here

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt (hunk 3 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/listener/ListenerSetParkingSpaces.kt:5]: because in the companion object of the Fragment when it's set the bundle, it use the .setSerializable to do it. So, its necessary to use the : Serializable

File: app/build.gradle (hunk 16 above)
Comment (florenciaferraiolo) [reply app/build.gradle:41]: sorry, I dont know when it got deleted. Done

File: app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/ParkingSizeUpdateDialogViewModel.kt (hunk 17 above)
Comment (florenciaferraiolo) [reply app/src/main/java/com/mobile/tandil/kotlinbaseproject/mvvm/viewmodel/ParkingSizeUpdateDialogViewModel.kt]: done

File: app/src/main/res/layout/fragment_parking_size_dialog.xml (hunk 18 above)
Comment (florenciaferraiolo) [reply app/src/main/res/layout/fragment_parking_size_dialog.xml]: done

File: app/src/main/res/values/dimen.xml (hunk 19 above)
Comment (florenciaferraiolo) [reply app/src/main/res/values/dimen.xml]: done

**Review by mauroLonghi — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]
